README.md
# Student Smart Printing Service 

[![formatter: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier) [![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/) ![CI](https://github.com/phucvinh57/fastify-template/actions/workflows/ci.yml/badge.svg) ![Deploy](https://github.com/phucvinh57/fastify-template/actions/workflows/release.yml/badge.svg)

## Prerequisites

- `docker` v20.10.22
- `docker-compose` v1.29.2
- `node` v18.13.0
- `npm` 8.19.3
- `cups` 2.4.1
- `cups-pdf` 3.0.1-14 (Virtual printer in case your computer is not connected to any actual printer)

## Commands

- `yarn bootstrap`: Set up development
- `yarn barrels`: Gather export objects from many files in a folder and re-export in `index.ts` file. See configs in `.barrelsby.json`.
- `yarn start`: Start application in dev mode
- `yarn db:migrate`: Apply new migration to current database
- `yarn db:reset`: Reset database and run seed
- `yarn db:deploy`: Deploy all migrations without confirmations (use in production)
- `yarn db:generate`: Just generate prisma client library
- `yarn db:studio`: Interact with database by a web UI
- `yarn lint`: Check linting
- `yarn format`: Format code
- `yarn start:docker`: Run `docker-compose.dev.yml` file to set up local database
- `yarn clean:docker`: Remove local database instance include its data.
- `yarn clean:git`: Clean local branches which were merged on remote

## Project structure

```py
📦prisma
 ┣ 📂migrations     # All SQL migration scripts go here
 ┣ 📜schema.prisma  # Database schema declaration
 ┗ 📜seed.ts        # Generate sample data script
📦src
 ┣ 📂configs        # Contain environment variables & other app configurations
 ┣ 📂constants      # Constants and enums go here
 ┣ 📂dtos           # Schema for input (from requests) & output (from responses)
 ┃ ┣ 📂in
 ┃ ┗ 📂out
 ┣ 📂handlers       # Handlers, which are responsible for handling core business logic
 ┣ 📂interfaces     # Interfaces
 ┣ 📂middlewares    # Middlewares such as logging or verifying tokens
 ┣ 📂plugins        # Plugin, in charge of organizing api routings & registering middleware
 ┣ 📂repositories   # Datasource configurations & connections. Could have more than one datasource.
 ┣ 📂services       # 3rd-party services or business logic services
 ┣ 📂types          # Types
 ┣ 📂utils          # Helping classes and functions
 ┣ 📜Server.ts      # Server setting & binding modules
 ┗ 📜index.ts       # Program entry
```


## Setup and Run Instructions

### Step 1: Set up Environment Variables
Create a .env file and fill it with the required environment variables. Ensure you replace placeholders like with the actual values.  
Note: Fill in `.env` file (use template from `.env.example`) before starts.  
For example:

``` .env
FASTIFY_PORT=3000
FASTIFY_TEST_PORT=3001

POSTGRES_USER=db_user
POSTGRES_PASSWORD=db_password
POSTGRES_DB=ssps_db
POSTGRES_PORT=5432
POSTGRES_TEST_PORT=5433
POSTGRES_URL=postgresql://db_user:db_password@localhost:5432/ssps_db

# ... (Other environment variables)
```

### Step 2: Run Bootstrap Command
Run the following command to set up the development environment:
```sh
yarn bootstrap
```
This command will likely install dependencies, set up configurations, and perform other necessary tasks.

### Step 3: Start the Application
Run the following command to start the application:
```sh
yarn start
```
This command will start your Node.js application. Ensure that ports 5432 and 8080 are available and not in use by other processes.

CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/TickLabVN/ssps-be/compare/v1.0.0...v2.0.0) (2023-12-07)


### ⚠ BREAKING CHANGES

* **paypal:** Switches the immediate unit of coin from USD to VND
* **execute printing request:** executing a printing request is now restricted to students only

### Features

* **coin:** add bonus coin for each 100,000 VND base on configurarion ([c53bef4](https://github.com/TickLabVN/ssps-be/commit/c53bef4bf247631e721a3c3e78a6f0c0b3dc72f7))
* **coin:** implement new handler and API dor bonusCoinPer100000VND configuration ([78b5f60](https://github.com/TickLabVN/ssps-be/commit/78b5f60630d4e29749084bd4d708b9c0de595d53))
* **coin:** implement new handler and API for coin to VND configuration ([b2edfad](https://github.com/TickLabVN/ssps-be/commit/b2edfad126b3f82934176ba8e746efb1eba173bb))
* **configuration printing:** save file after approving configuration on Minio for preview ([f3a0a35](https://github.com/TickLabVN/ssps-be/commit/f3a0a35c9a8a8389225d8bff26da237f45bfa296))
* **configuration:** implement serving fee configuration and missing get configuration APIs ([456d240](https://github.com/TickLabVN/ssps-be/commit/456d2404713c51e2328b594d67349cb45e048d5b))


### Bug Fixes

* **coin:** convert from amount VND to USD rounded to two decimal places ([8466bcf](https://github.com/TickLabVN/ssps-be/commit/8466bcfd4b23d8b17868b76557c3ea043028149d))
* **coin:** round VND conversion from USD to avoid missing bonus coin ([4f8d8f3](https://github.com/TickLabVN/ssps-be/commit/4f8d8f32370bae769a282879ca6efad47bb49247))
* **execute printing request:** decrease appropriate user's remain coins base on printing request ([83074d2](https://github.com/TickLabVN/ssps-be/commit/83074d279e1176cc5a97c49cfdeafb8b7bd08079))
* **permission:** change permission of get configuration APIs to be accessible by all roles ([56b1ac6](https://github.com/TickLabVN/ssps-be/commit/56b1ac63b3c0facefa8a3037c922653c387b2fb1))
* **printer:** let CUPS choose a printer ([7073e3d](https://github.com/TickLabVN/ssps-be/commit/7073e3d160156ff2b3542cd6f35ac9b40dba0b44))
* **printing request:** adjust coins and pages with changes in amount of pages when config changes ([6968c22](https://github.com/TickLabVN/ssps-be/commit/6968c22a8282e91a0e6767b9426a034d30afc596))
* **printing:** change synchronize handler of execute printing request to use a for loop ([68495ae](https://github.com/TickLabVN/ssps-be/commit/68495ae7d26dabe8d964113923561c02aaa8ec9b))
* **printing:** send only buffer to printer in development environment ([7d0a346](https://github.com/TickLabVN/ssps-be/commit/7d0a346de5a7ce4d5100084877cf1cee01f3f19a))


### Code Refactoring

* **paypal:** change amount of create PayPal Order to VND; use VND as immediate unit of coin ([cd79516](https://github.com/TickLabVN/ssps-be/commit/cd79516923eb215497768d6bbf96558dc26c72c5))

## 1.0.0 (2023-12-03)


### ⚠ BREAKING CHANGES

* **env.ts:** Changing to using process.env instead of envs (envalid). Need to fix later.
* **printing request:** change from node-thermal-printer to node-printer

### Features

* **auth:** implement API for logout ([1143d44](https://github.com/TickLabVN/ssps-be/commit/1143d44b59c4bf58b2485fd20b0094029d845fa1))
* **auth:** implement authorization with admin and student ([565a308](https://github.com/TickLabVN/ssps-be/commit/565a30819109650995fb3146179bcd7596409303))
* **cd:** implement CD for cloned containers to ensure the success of development deployments ([c0d9f07](https://github.com/TickLabVN/ssps-be/commit/c0d9f07f9a40ff0abbe244361053822b791cda12))
* **configuration:** create new table for saving configurations ([8ca0d88](https://github.com/TickLabVN/ssps-be/commit/8ca0d88080fe12dd1130fb34e8ec3477bd617d31))
* **configuration:** implement API for getting all configurations and updating accepted extensions ([ac726ec](https://github.com/TickLabVN/ssps-be/commit/ac726ec7f5c4e98489f5eeedf004fbe9fb10562e))
* **configuration:** move dollar to coin config to db ([6416e91](https://github.com/TickLabVN/ssps-be/commit/6416e91fe257a82c27e926c645c2ba6d5b2e1dc7))
* **dev-dependency:** add commitlint and commitizen for conventional commit ([413d4f1](https://github.com/TickLabVN/ssps-be/commit/413d4f10ee2efdc09c7f44242108e0d6f0f901ac))
* **docker-compose:** modify docker-compose to using .env file and using ssps external network ([2040e31](https://github.com/TickLabVN/ssps-be/commit/2040e315fdf6ba2680d82c57eae64094c65c0ee0))
* **edit pdf file:** uncomple edit pdf file for config printing file ([e37593f](https://github.com/TickLabVN/ssps-be/commit/e37593f6d7ae4c948ce50897a64aa84c86d6e8f8))
* **edit pdf:** change select page custom from array to comma-separated string ([46a03ff](https://github.com/TickLabVN/ssps-be/commit/46a03fffc1a497882cded2c82b21b1e95270507f))
* **edit pdf:** implement a comprehensive function wrapping the entire setting functionality ([516be6a](https://github.com/TickLabVN/ssps-be/commit/516be6a389b2ee36d2470c46a67f5cb3bbb4b03d))
* **edit pdf:** implement landscape with page per sheet parameter ([0c01474](https://github.com/TickLabVN/ssps-be/commit/0c01474ed9105f4113c286d81e1205c2eef1050f))
* **edit pdf:** implement portrait with page per sheet parameter ([7d7ca4b](https://github.com/TickLabVN/ssps-be/commit/7d7ca4bf4791f38552b5c3d093c4e5a10b615732))
* **edit pdf:** implement setting for two-sided short or long edge ([f4ea167](https://github.com/TickLabVN/ssps-be/commit/f4ea167255f172926cdcd69a10d15882352d15ec))
* **google oauth:** implement handler logic for google oauth callback ([3e1c192](https://github.com/TickLabVN/ssps-be/commit/3e1c192db52ba2f971043a5286fef56208224159))
* **home page:** implement an api to send slide image urls of home page ([a37c900](https://github.com/TickLabVN/ssps-be/commit/a37c900218c4b378f5b1bbc5fa443bfc7b988c26))
* **printing request:** change to use node-printer, implement logic print whole printing request ([07fbc16](https://github.com/TickLabVN/ssps-be/commit/07fbc16551c09e44db93ace500f7598c2cc4d25a))

prisma/seed.ts
import { hashSync } from 'bcrypt';
import { PrismaClient, Student } from '@prisma/client';

const USER_ROLES = {
    admin: 5150,
    student: 2001,
    printer_manager: 3001,
    maintainer: 4001
};

type UserRole = keyof typeof USER_ROLES;

const prisma = new PrismaClient();
const SALT_ROUNDS = 10;

const createUser = async () => {
    const users: { userName: string; password: string; roles: UserRole[]; name: string; email: string }[] = [
        {
            userName: 'student@example.com',
            password: '123456789',
            roles: ['student'],
            name: 'Student name',
            email: 'student@example.com'
        },
        {
            userName: 'ssps@example.com',
            password: '123456789',
            roles: ['admin'],
            name: 'Admin name',
            email: 'admin@example.com'
        },
        {
            userName: 'root@example.com',
            password: '123456789',
            roles: ['admin', 'student'],
            name: 'Root name',
            email: 'root@example.com'
        }
    ];

    const handledUsers = users.map((user) => {
        const { password, roles, ...remain } = user;
        const hashPassword = hashSync(password, SALT_ROUNDS);
        const roleValue: number[] = roles.map((role) => USER_ROLES[role]);
        return {
            ...remain,
            password: hashPassword,
            role: roleValue
        };
    });

    const sampleUsers = await prisma.user.createMany({
        data: handledUsers
    });

    console.log(sampleUsers);
};

const createStudent = async () => {
    // Fetch the user ID for a student with the specified role
    const studentUser = await prisma.user.findMany({
        select: {
            id: true
        },
        where: {
            role: { has: USER_ROLES.student }
        }
    });

    // Check if a student user with the specified role exists
    if (!studentUser) {
        console.error('No student user found with the specified role.');
        return;
    }

    const students: Student[] = studentUser.map((user) => {
        return {
            default_coin_per_sem: 100,
            remain_coin: 50,
            id: user.id
        };
    });

    const sampleStudents = await prisma.student.createMany({
        data: students
    });

    console.log(sampleStudents);
};

const createLocation = async () => {
    const locations: { address: string }[] = [
        {
            address: 'H1'
        },
        {
            address: 'H2'
        },
        {
            address: 'H3'
        },
        {
            address: 'H6'
        }
    ];

    const sampleLocations = await prisma.location.createMany({
        data: locations
    });

    console.log(sampleLocations);
};

// function getRandomElementFromArray<T>(arr: T[]): T | undefined {
//     if (arr.length === 0) {
//         return undefined;
//     }

//     const randomIndex = Math.floor(Math.random() * arr.length);
//     return arr[randomIndex];
// }

// const createPrintingRequest = async () => {
//     const studentUser = await prisma.user.findFirst({
//         select: {
//             id: true
//         },
//         where: {
//             role: { has: USER_ROLES.student }
//         }
//     });

//     const locations = await prisma.location.findMany({
//         select: { id: true }
//     });

//     if (!studentUser) {
//         console.error('No student user found with the specified role.');
//         return;
//     }

//     const printingRequests: {
//         status: $Enums.PrintingStatus;
//         locationId: string | undefined;
//         numFiles: number;
//         numPages: number;
//         coins: number;
//         paid: $Enums.Paid;
//         userId: string;
//     }[] = [
//         {
//             status: 'progressing',
//             locationId: getRandomElementFromArray(locations)?.id,
//             numFiles: 0,
//             numPages: 0,
//             coins: 0,
//             paid: 'not_paid',
//             userId: studentUser.id
//         },
//         {
//             status: 'ready',
//             locationId: getRandomElementFromArray(locations)?.id,
//             numFiles: 0,
//             numPages: 0,
//             coins: 0,
//             paid: 'not_paid',
//             userId: studentUser.id
//         },
//         {
//             status: 'done',
//             locationId: getRandomElementFromArray(locations)?.id,
//             numFiles: 0,
//             numPages: 0,
//             coins: 0,
//             paid: 'not_paid',
//             userId: studentUser.id
//         },
//         {
//             status: 'canceled',
//             locationId: getRandomElementFromArray(locations)?.id,
//             numFiles: 0,
//             numPages: 0,
//             coins: 0,
//             paid: 'not_paid',
//             userId: studentUser.id
//         },
//         {
//             status: 'done',
//             locationId: getRandomElementFromArray(locations)?.id,
//             numFiles: 0,
//             numPages: 0,
//             coins: 0,
//             paid: 'not_paid',
//             userId: studentUser.id
//         }
//     ];

//     const samplePrintingRequest = await prisma.printingRequest.createMany({
//         data: printingRequests
//     });

//     console.log(samplePrintingRequest);
// };

const createConfiguration = async () => {
    const acceptedExtensions = ['pdf', 'png'];
    const configuration: {
        name: string;
        value: string;
        description: string;
    }[] = [
        { name: 'coin per page', value: '2', description: 'The amount of coin a student needs to print one page' },
        { name: 'dollar to coin', value: '73', description: 'The amount of coin user gets per dollar' },
        { name: 'coin to vnd', value: '200', description: 'The value of one coin exchange to vnd' },
        { name: 'bonus coin per 100000 vnd', value: '20', description: 'The amount of coin is bonus for 100 000 vnd' },
        { name: 'coin per sem', value: '100', description: 'The amount of coin a student has free in one semester' },
        ///100mb = 100 * 1024 * 1024 (byte)
        { name: 'max file size', value: `${100 * 1024 * 1024}`, description: 'The amount of coin a student has free in one semester' },
        { name: 'service fee', value: `5`, description: 'The amount of coin for a printing request be executed' }
    ];

    const serializedExtensions = JSON.stringify(acceptedExtensions);

    configuration.push({ name: 'accepted extensions', value: serializedExtensions, description: 'Supported file extensions of printer' });

    const sampleConfiguration = await prisma.configuration.createMany({ data: configuration });

    console.log(sampleConfiguration);
};

async function generateSampleData() {
    await createUser();

    await createStudent();

    await createLocation();

    await createConfiguration();

    process.exit(0);
}

generateSampleData();
